### PR TITLE
Add `get_upstream_resource` to resource adaptors

### LINKED ISSUE #1455: Add `get_upstream_resource` method to resource adaptors to eventually replace raw `get_upstream`

### PR DESCRIPTION
**Pull request by miscco**
We want to get rid of the raw `get_upstream` method. In order to achieve that we need to add an alternative that is based on our new `rmm::device_async_resource_ref`

I am a stickler to consistency so I also cleaned the documentation of `get_upstream` up and added `[[nodiscard]]`

closes #1455

### CODE REVIEW

File: include/rmm/mr/device/binning_memory_resource.hpp
Comment (harrism) [inline include/rmm/mr/device/binning_memory_resource.hpp:107]: non-blocking question: I wonder if we should make the naming of upstream resources consistent. Seems about 50/50 `upstream_` and `upstream_mr`

**Review by harrism — APPROVED** [review]
Looks good, thank you.  Should we standardize the interface of resource adaptors somehow?  I've often felt that MRs would be more composable if there were some common interface for accessing upstream MRs.

File: include/rmm/mr/device/binning_memory_resource.hpp
Comment (miscco) [reply include/rmm/mr/device/binning_memory_resource.hpp:107]: Oh yeah, that rubbed me too, but I thought that it would be a bit to far away from the purpose of this PR

**Comment by miscco** [comment]
> Looks good, thank you. Should we standardize the interface of resource adaptors somehow? I've often felt that MRs would be more composable if there were some common interface for accessing upstream MRs.

I believe we are on a good way. When moving everything towards `async_device_memory_resource` we can completely get rid of `Upstream` altogether and only over communicate throught the `resource_ref`

**Comment by harrism** [comment]
But adaptors, suballocators, etc. will always have upstream resource refs. Just wondering if we need a legislated way to get the ref, or if we just stick with a de facto interface.

**Comment by miscco** [comment]
> But adaptors, suballocators, etc. will always have upstream resource refs. Just wondering if we need a legislated way to get the ref, or if we just stick with a de facto interface.

I mean my thought was that we will just keep going with `get_upstream_resource` or `get_upstream`

**Comment by jrhemstad** [comment]
> I mean my thought was that we will just keep going with `get_upstream_resource` or `get_upstream`

I believe @harrism is asking about adding a `resource_adaptor` concept and/or CRTP helper that codifies `get_upstream_resource()` as the definitive way to access an upstream resource in an adaptor.

**Comment by miscco** [comment]
Coming back to this, I am all for providing a well defined way of extracting the upstream resource 

However, we want to make sure that we have the right one. Currently we have `get_upstream` and this PR introduces the slightly longer `get_upstream_resource`

If we want to introduce a concept that checks whether an upstream resource is provided we need to decide whether we want `get_upstream_resource` going forward or whether we want to move back to `get_upstream` once the old one has been removed

**Comment by harrism** [comment]
My 2c: I think `get_upstream_resource` is very clearly named.  So it would be fine to switch to that if we have to change the API from pointers to refs anyway.

File: include/rmm/mr/is_resource_adaptor.hpp
Comment (miscco) [reply include/rmm/mr/is_resource_adaptor.hpp:39]: This checks whether something has a `get_upstream_resource()` method.

If so it then verifies that Resource satisfies `cuda::mr::resource<Resource>`

clang-format is lifting a lot here with the formatting. Not that at the end of line 38 there is a =

So the short form is `inline constexpr bool is_resource_adaptor<Resource, void_t> = cuda::mr::resource<Resource>`

File: include/rmm/mr/device/thrust_allocator_adaptor.hpp
Comment (miscco) [inline include/rmm/mr/device/thrust_allocator_adaptor.hpp:124]: This has just been added in the recent push so I am pretty sure it is safe to rename the method so that it is consistent with the other adaptors

File: include/rmm/mr/device/pool_memory_resource.hpp
Comment (miscco) [inline include/rmm/mr/device/pool_memory_resource.hpp:235]: This has just been added in the recent push so I am pretty sure it is safe to rename the method so that it is consistent with the other adaptors

File: include/rmm/mr/is_resource_adaptor.hpp
Comment (harrism) [reply include/rmm/mr/is_resource_adaptor.hpp:40]: I don't understand how assignment works in this situation.

Comment (harrism) [reply include/rmm/mr/is_resource_adaptor.hpp:40]: Oh I get it. If you get a match, then this returns true if the type is a resource.

File: include/rmm/mr/device/pool_memory_resource.hpp
Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:355]: What PR added it?

Comment (miscco) [reply include/rmm/mr/device/pool_memory_resource.hpp:355]: That was my initial memory_resource PR: https://github.com/rapidsai/rmm/pull/1095

Comment (harrism) [reply include/rmm/mr/device/pool_memory_resource.hpp:355]: Ah, thought you meant someone else had added it.

**Comment by harrism** [comment]
I generally like to get 2 C++ codeowner approvals before merging.  Just ping the auto-assigned reviewers if it has been a while.

**Pull request merged**
